test/source/dense_test.cpp

```diff
@@ -21,6 +21,9 @@ TEST_CASE("Dense-BaseCase")
         ++loop;
     }
     CHECK(loop == SIZE);
+
+    // Check basic properties of all dense levels
+    CHECK(d.LevelProperty().is_full);
```

Comment:
@hameerabbasi and @bharath2438 is this sort of what we would want API wise?

Reply:
Yeah something like this or `d::LevelProperties::is_full` since we want it to be a type

Reply:
> Yeah something like this or `d::LevelProperties::is_full` since we want it to be a type

Wouldn't that one be a bit more difficult since that's accessing the namespace of the class of `d`, whereas, we wouldn't necessarily "know" the class object, so I took the route of using a `constexpr` function to access the internal `LevelProperties` attribute for each level class.

Reply:
Dense is strictly ordered as well, but otherwise your statemen holds.

Reply:
I noticed that the strictly ordered is not required in `dense`. I added it now.

Reply:
Wait actually, this does not work as `TEST_CASE("Dense-2D")` inside `dense_test.cpp` fails when we add `static_assert(LevelProperties::is_ordered);` to the `dense` class.

Reply:
Hmm, that's weird. Dense absolutely cannot be unordered, the unit test must be faulty.

Reply:
Okay, I modified it in https://github.com/hameerabbasi/xsparse/pull/22/commits/2ac6d71451ea7a9212d2d33cdd87bb1717b236f9

Reply:
![image](https://github.com/hameerabbasi/xsparse/assets/3460267/f399d2e4-d507-456b-8c3d-d6f89dc45417)

It seems this table from the PhD thesis suggests it doesn't have to be ordered. But I don't understand the intricacies there...

Reply:
It's weird since `Dense` is the only level that will always be iterated in order regardless of how you insert into it, so that doesn't make sense to me.

Reply:
> It's weird since `Dense` is the only level that will always be iterated in order regardless of how you insert into it, so that doesn't make sense to me.

Yes even I was confused about this. Also, shouldn't dense strictly be unique, since it encodes all coordinates in the range N?